st2common/st2common/util/config_loader.py

```diff
@@ -112,9 +114,56 @@ def _get_object_property_schema(object_schema, additional_properties_keys=None):
             # ensure that these keys are present in the object
             for key in additional_properties_keys:
                 property_schema[key] = additional_properties
-        property_schema.update(object_schema.get("properties", {}))
+
+        properties_schema = object_schema.get("properties", {})
+        property_schema.update(properties_schema)
+
+        potential_patterned_keys = set(additional_properties_keys) - set(
```

Comment:
Maybe a comment, it took me a while to double-check this. But I think your logic is ok.

additionalProperties only applies to keys that are not matched by properties or patternProperties. So patternProperties takes precedence. Therefore any keys that have their own schema AND match the patternProperties are going to have the pattern rule applied rather than the additionalProperty defined.

I can see this confusing some, so I wonder as well if we should have a debug log statement for any keys that do get their property updated.

Reply:
No, this is calculating which keys are not explicitly in `properties`. Then, patternProperties is only compared against this subset of keys.

I can see hope this code is confusing. I just added some comments and renamed a few variables to hopefully make this more grokkable. Could you take another look?

Do my changes clarify your question, or did I misunderstand?